### ISSUE TITLE
pcolormesh edges do not follow lat/lon lines

### ISSUE THREAD

**Issue by AgilentGCMS** [SciTools/cartopy#1968]
### Description
I would like to plot a gridded field on a map with pcolormesh. However, the colored lat/lon boxes do not have edges that correspond to the lat/lon values. Here's a MWE with a very coarse resolution field to highlight the problem:

```python
from matplotlib import pyplot as plt
import numpy as np
import cartopy.crs as ccrs
import cartopy.feature as cfeature

# set resolution of plotted field
nlat = 6
nlon = 12
# generate a random field and the lat and lon edges
plot_field = np.random.standard_normal((nlat, nlon))
lat_edges = np.linspace(-90., 90., nlat+1)
lon_edges = np.linspace(-180., 180., nlon+1)

map_proj = ccrs.Robinson(central_longitude=0.0)
data_proj = ccrs.PlateCarree(central_longitude=0.0)

# make sure that the map covers the entire plot area
aspect = (map_proj.x_limits[1]-map_proj.x_limits[0])/(map_proj.y_limits[1]-map_proj.y_limits[0])
height = 5.0
width = aspect * height
fig = plt.figure(figsize=(width, height))
ax = plt.axes([0.0, 0.0, 1.0, 1.0], projection=map_proj)
ax.add_feature(cfeature.COASTLINE.with_scale('50m'), linewidth=0.7, edgecolor='k')
ax.add_feature(cfeature.LAKES.with_scale('50m'), linewidth=0.5, edgecolor='k', facecolor='none')
ax.gridlines(xlocs=np.arange(-180.,181.,30.), ylocs=np.arange(-90.,91.,30.), linewidth=0.5, linestyle=(0,(5,2)), color='0.25')

img = ax.pcolormesh(lon_edges, lat_edges, plot_field, transform=data_proj, cmap=plt.cm.Spectral, vmin=-1.0, vmax=1.0, edgecolors='face', shading='flat')

plt.show()
```

The [resulting plot](https://drive.google.com/file/d/1IfWxNxu9_IlqkdAFMjxS_tGQOKenDRr0/view?usp=sharing) shows two problems:

1. The longitude edges of the colored faces are not geodesics, they're straight lines in plot coordinates. This is especially noticeable at the eastern and western edges, but is visible everywhere. I'd like a way to make them geodesics.
2. The latitude edges of the colored faces do not correspond to the latitude boundaries. Either the latitudes are drawn at the wrong locations, or the face edges.

### Setup details

* Operating system: MacOS
* Cartopy version: 0.18.0
* Python version: 3.8.12
* Matplotlib version: 3.4.3
* Numpy version: 1.21.2

**Comment by greglucas** [comment]
1. `pcolormesh()` is meant to be fast and doesn't transform the paths between the corners defining the quadrilaterals, only the corner points themselves. If you need higher resolution you can switch to using `pcolor()` which produces the desired curves by transforming points along each path, it is just much slower.
2. Could you clarify? It looks like the latitudes of the edges of the boxes are lined up with the gridlines.

**Comment by AgilentGCMS** [comment]
1. OK, I get your point about pcolormesh being a fast method. Using pcolor I indeed get the edges to follow geodesics. Thanks!
2. No, they're not, although they're close. In [this version](https://drive.google.com/file/d/1f0Rk-bHoOxRQhUv9hGAkyJ_F60buXFsy/view?usp=sharing) I've circled a few edges where the problem is most obvious. The latitude lines should pass through the interfaces between two colors. Instead, they're entirely within one color.
